eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala

```diff
@@ -276,8 +276,10 @@ object Commitments {
 
     // we allowed mismatches between our feerates and our remote's as long as commitments didn't contain any HTLC at risk
     // we need to verify that we're not disagreeing on feerates anymore before offering new HTLCs
-    val localFeeratePerKw = feeConf.feeEstimator.getFeeratePerKw(target = feeConf.feeTargets.commitmentBlockTarget)
-    if (Helpers.isFeeDiffTooHigh(localFeeratePerKw, commitments.localCommit.spec.feeratePerKw, feeConf.maxFeerateMismatchFor(commitments.remoteNodeId))) {
+    // NB: there may be a pending update_fee that hasn't been signed yet that needs to be taken into account
+    val currentFeeratePerKw = commitments.remoteChanges.proposed.collect { case f: UpdateFee => f.feeratePerKw }.lastOption.getOrElse(commitments.localCommit.spec.feeratePerKw)
```

Comment:
We should use `remoteChanges.acked ++ remoteChanges.signed` instead, because we haven't yet acknowledged what is in `remoteChanges.proposed`.

Reply:
I tried moving that check in `sendCommit`, but it's not working well, because a failure in `sendCommit` only fails the command but doesn't close the channel. If that happens the channel will just be kinda stuck, but not clearly closed.

So I decided in https://github.com/ACINQ/eclair/pull/1672/commits/0602beefd487eda7148a7741d43aeae95cf0bbbd to be quite aggressive and check all pending `update_fee` from our remote. Maybe there are some we could ignore because they would be overridden by later ones, but I don't want to take any risk by adding complexity here.

The advantage of checking this in `sendAdd` / `receiveAdd` is that in `sendAdd`, we simply reject the HTLC (without closing the channel, we wait for the channel to be back at a feerate we agree with before using it) and only in `receiveAdd` we close the channel. That sounds like a good compromise, let me know what you think.

Reply:
I think a better idea would be to use `CommitmentSpec.reduce` exactly like in the `receiveFee` method, it would take care of deciding which fee to take into account and be consistent with the rest.

Reply:
I'm afraid that wouldn't be enough, there are still issues when we add an HTLC after letting some `update_fee`s we disagree with slip into the `remoteChanges`.

Imagine there is no pending HTLC but there is an `update_fee` in `remoteChanges.proposed` and one in `remoteChanges.acked`, and we're disagree only with the one in `remoteChanges.proposed`. Then the remote adds an HTLC.

If we only check against the `update_fee` in `remoteChanges.signed`, the `update_fee` in `remoteChanges.proposed` will get signed at the same time as the HTLC and we end up with a signed commitment, with a pending HTLC and a feerate we disagree with.

If on the other hand we agree with the `update_fee` in `remoteChanges.proposed` but disagree with the one in `remoteChanges.signed`, but this time we add an HTLC (not the remote). They can delay the ack/sign process on the `update_fee` in `remoteChanges.proposed` and we can end up with a commitment with our HTLC and the feerate we disagree with.

It's really hard to explore all the potential cases since the protocol is asynchronous, and I'm really not feeling safe leaving a hole there, so I'd rather check all pending `update_fee`s and close if we disagree with any of them, this way we're sure our HTLCs are safe. Does that make sense?

Reply:
I agree, my proposal doesn't work.

After reading your comment I immediately went to check if we weren't exposed to the opposite scenario, where we first send an `update_add_htlc`, and then received an `update_fee` that we disagree with. But we _do_ check all (unsigned and signed, incoming and outgoing) htlcs in `hasPendingOrProposedHtlcs` so we are okay. But it doesn't make me feel safe either.

This makes me go back to the original idea of enforcing a check in `sendCommit`, which is really the only way to be sure. Going back to your comment https://github.com/ACINQ/eclair/pull/1672#discussion_r577466335:

> I tried moving that check in sendCommit, but it's not working well, because a failure in sendCommit only fails the command but doesn't close the channel. If that happens the channel will just be kinda stuck, but not clearly closed.

We could define a `ChannelException` that makes us force close the channel.

> The advantage of checking this in sendAdd / receiveAdd is that in sendAdd, we simply reject the HTLC (without closing the channel, we wait for the channel to be back at a feerate we agree with before using it) and only in receiveAdd we close the channel. That sounds like a good compromise, let me know what you think.

That is a really good point. Sigh... But maybe we should do both? Our actual security check would be in `sendCommit`, and we would have a best-effort check in `sendAdd` to prevent unnecessary force closes. What do you think?

Reply:
> We could define a ChannelException that makes us force close the channel.

We could indeed, but that breaks the symmetry between commands and external messages and feels a bit hackish...

> But maybe we should do both? Our actual security check would be in sendCommit, and we would have a best-effort check in sendAdd to prevent unnecessary force closes. What do you think?

I think that if we have this check in `sendAdd`, we should have it in `receiveAdd` for consistency (when we read the code later, we'll wonder why it's not in both, even with a comment explaining it since this is quite subtle). And in that case the check in `sendCommit` is unnecessary! What do you dislike about checking `proposed + acked + signed` here? It's quite concise and we're sure it does the job, doesn't it?

Reply:
> It's quite concise and we're sure it does the job, doesn't it?

That's the thing. I _think_ it works, but I'm not 100% sure it's enough. Like you said, it's really hard to explore all potential cases.

Anyway, there are a lot of other things that are even more sensitive than fee rate, that we don't double check in the `sendCommit` (e.g. expiries), so it wouldn't really make sense to do a special case here. Let's forget about it.